substrate/client/authority-discovery/src/worker/addr_cache.rs

```diff
@@ -172,8 +240,224 @@ fn addresses_to_peer_ids(addresses: &HashSet<Multiaddr>) -> HashSet<PeerId> {
+	}
+}
+
+/// A (de)serializable version of [`PeerId`] that can be used for serialization,
+#[serde_as]
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq, Hash)]
+#[serde(transparent)]
+struct SerializablePeerId {
+	#[serde_as(as = "serde_with::hex::Hex")]
+	bytes: Vec<u8>,
+}
+
+impl From<PeerId> for SerializablePeerId {
+	fn from(peer_id: PeerId) -> Self {
+		Self { bytes: peer_id.to_bytes() }
+	}
+}
+impl TryFrom<SerializablePeerId> for PeerId {
+	type Error = Error;
+
+	fn try_from(value: SerializablePeerId) -> Result<Self, Self::Error> {
+		PeerId::from_bytes(&value.bytes)
+			.map_err(|e| Error::EncodingDecodingAddrCache(e.to_string()))
+	}
+}
+
+/// A (de)serializable version of [`Multiaddr`] that can be used for serialization,
+#[serde_as]
+#[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
+#[serde(transparent)]
+struct SerializableMultiaddr {
+	/// `Multiaddr` holds a single `LiteP2pMultiaddr`, which holds `Arc<Vec<u8>>`.
+	#[serde_as(as = "serde_with::hex::Hex")]
+	bytes: Vec<u8>,
+}
+impl From<Multiaddr> for SerializableMultiaddr {
+	fn from(multiaddr: Multiaddr) -> Self {
+		Self { bytes: multiaddr.to_vec() }
+	}
+}
+impl TryFrom<SerializableMultiaddr> for Multiaddr {
+	type Error = ParseError;
+
+	fn try_from(value: SerializableMultiaddr) -> Result<Self, Self::Error> {
+		Self::try_from(value.bytes)
+	}
+}
+
+/// Writes the content to a file at the specified path.
+fn write_to_file(path: &PathBuf, content: &str) -> io::Result<()> {
+	let mut file = File::create(path)?;
+	file.write_all(content.as_bytes())?;
+	file.flush()
+}
+
+/// Reads content from a file at the specified path and tries to JSON deserializes
+/// it into the specified type.
+fn load_from_file<T: DeserializeOwned>(path: &PathBuf) -> io::Result<T> {
+	let file = File::open(path)?;
+	let reader = BufReader::new(file);
+
+	serde_json::from_reader(reader).map_err(|e| {
+		error!(target: super::LOG_TARGET, "Failed to load from file: {}", e);
+		io::Error::new(io::ErrorKind::InvalidData, e)
+	})
+}
+
+/// Asynchronously writes content to a file using a background thread.
+/// Multiple consecutive writes in quick succession (before the current write is completed)
+/// will be **throttled**, and only the last write will be performed.
+#[derive(Clone)]
+pub struct ThrottlingAsyncFileWriter {
+	/// Each request to write content will send a message to this sender.
+	///
+	/// N.B. this is not passed in as an argument, it is an implementation
+	/// detail.
+	sender: mpsc::UnboundedSender<String>,
+}
+
+impl ThrottlingAsyncFileWriter {
+	/// Creates a new `ThrottlingAsyncFileWriter` for the specified file path,
+	/// the label is used for logging purposes.
+	pub fn new(purpose_label: String, path: impl Into<PathBuf>) -> Self {
+		let path = Arc::new(path.into());
+		let (sender, mut receiver) = mpsc::unbounded();
+
+		let path_clone = Arc::clone(&path);
+		thread::spawn(move || {
```

Comment:
I would pass a SpawnHandle from here: https://github.com/paritytech/polkadot-sdk/blob/b9fbf243c57939ecadc89b82ed42249703203874/polkadot/node/service/src/builder/mod.rs#L573.


And use spawn_handle.spawn_blocking, it just makes the task more visible to the monitoring.

Reply:
you mean pass the `task_manager.spawn_handle()`? from https://github.com/paritytech/polkadot-sdk/blob/b9fbf243c57939ecadc89b82ed42249703203874/polkadot/node/service/src/builder/mod.rs#L99 ?